File: agenda_backend/lib/core/models/recurrence_rule.dart
/// Modello per le regole di ricorrenza delle prenotazioni
class RecurrenceRule {
  final int id;
  final int businessId;
  final RecurrenceFrequency frequency;
  final int intervalValue;
  final int? maxOccurrences;
  final DateTime? endDate;
  final ConflictStrategy conflictStrategy;
  final List<int>? daysOfWeek;
  final int? dayOfMonth;
  final DateTime createdAt;
  final DateTime updatedAt;

  const RecurrenceRule({
    required this.id,
    required this.businessId,
    required this.frequency,
    this.intervalValue = 1,
    this.maxOccurrences,
    this.endDate,
    this.conflictStrategy = ConflictStrategy.skip,
    this.daysOfWeek,
    this.dayOfMonth,
    required this.createdAt,
    required this.updatedAt,
  });

  factory RecurrenceRule.fromJson(Map<String, dynamic> json) {
    return RecurrenceRule(
      id: json['id'] as int,
      businessId: json['business_id'] as int,
      frequency: RecurrenceFrequency.fromString(json['frequency'] as String),
      intervalValue: json['interval_value'] as int? ?? 1,
      maxOccurrences: json['max_occurrences'] as int?,
      endDate: json['end_date'] != null
          ? DateTime.parse(json['end_date'] as String)
          : null,
      conflictStrategy: ConflictStrategy.fromString(
        json['conflict_strategy'] as String? ?? 'skip',
      ),
      daysOfWeek: json['days_of_week'] != null
          ? List<int>.from(json['days_of_week'] as List)
          : null,
      dayOfMonth: json['day_of_month'] as int?,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'business_id': businessId,
    'frequency': frequency.value,
    'interval_value': intervalValue,
    if (maxOccurrences != null) 'max_occurrences': maxOccurrences,
    if (endDate != null) 'end_date': endDate!.toIso8601String().split('T')[0],
    'conflict_strategy': conflictStrategy.value,
    if (daysOfWeek != null) 'days_of_week': daysOfWeek,
    if (dayOfMonth != null) 'day_of_month': dayOfMonth,
    'created_at': createdAt.toIso8601String(),
    'updated_at': updatedAt.toIso8601String(),
  };

  RecurrenceRule copyWith({
    int? id,
    int? businessId,
    RecurrenceFrequency? frequency,
    int? intervalValue,
    int? maxOccurrences,
    DateTime? endDate,
    ConflictStrategy? conflictStrategy,
    List<int>? daysOfWeek,
    int? dayOfMonth,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return RecurrenceRule(
      id: id ?? this.id,
      businessId: businessId ?? this.businessId,
      frequency: frequency ?? this.frequency,
      intervalValue: intervalValue ?? this.intervalValue,
      maxOccurrences: maxOccurrences ?? this.maxOccurrences,
      endDate: endDate ?? this.endDate,
      conflictStrategy: conflictStrategy ?? this.conflictStrategy,
      daysOfWeek: daysOfWeek ?? this.daysOfWeek,
      dayOfMonth: dayOfMonth ?? this.dayOfMonth,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  /// Descrizione leggibile della ricorrenza (es. "Ogni 2 settimane")
  String toReadableString({String locale = 'it'}) {
    final interval = intervalValue;
    switch (frequency) {
      case RecurrenceFrequency.daily:
        if (interval == 1) {
          return locale == 'it' ? 'Ogni giorno' : 'Every day';
        }
        return locale == 'it'
            ? 'Ogni $interval giorni'
            : 'Every $interval days';
      case RecurrenceFrequency.weekly:
        if (interval == 1) {
          return locale == 'it' ? 'Ogni settimana' : 'Every week';
        }
        return locale == 'it'
            ? 'Ogni $interval settimane'
            : 'Every $interval weeks';
      case RecurrenceFrequency.monthly:
        if (interval == 1) {
          return locale == 'it' ? 'Ogni mese' : 'Every month';
        }
        return locale == 'it'
            ? 'Ogni $interval mesi'
            : 'Every $interval months';
      case RecurrenceFrequency.custom:
        return locale == 'it'
            ? 'Ogni $interval giorni'
            : 'Every $interval days';
    }
  }
}

/// Frequenza di ricorrenza
enum RecurrenceFrequency {
  daily('daily'),
  weekly('weekly'),
  monthly('monthly'),
  custom('custom');

  final String value;
  const RecurrenceFrequency(this.value);

  static RecurrenceFrequency fromString(String value) {
    return RecurrenceFrequency.values.firstWhere(
      (e) => e.value == value,
      orElse: () => RecurrenceFrequency.weekly,
    );
  }
}

/// Strategia per gestione conflitti
enum ConflictStrategy {
  skip('skip'),
  force('force');

  final String value;
  const ConflictStrategy(this.value);

  static ConflictStrategy fromString(String value) {
    return ConflictStrategy.values.firstWhere(
      (e) => e.value == value,
      orElse: () => ConflictStrategy.skip,
    );
  }
}

/// Configurazione per creare una nuova ricorrenza (lato UI)
class RecurrenceConfig {
  final RecurrenceFrequency frequency;
  final int intervalValue;
  final int? maxOccurrences;
  final DateTime? endDate;
  final ConflictStrategy conflictStrategy;

  const RecurrenceConfig({
    this.frequency = RecurrenceFrequency.weekly,
    this.intervalValue = 1,
    this.maxOccurrences,
    this.endDate,
    this.conflictStrategy = ConflictStrategy.skip,
  });

  /// Indica se la ricorrenza ha un termine definito
  bool get hasEnd => maxOccurrences != null || endDate != null;

  /// Converte in payload per API
  Map<String, dynamic> toApiPayload() => {
    'frequency': frequency.value,
    'interval_value': intervalValue,
    if (maxOccurrences != null) 'max_occurrences': maxOccurrences,
    if (endDate != null) 'end_date': endDate!.toIso8601String().split('T')[0],
    'conflict_strategy': conflictStrategy.value,
  };

  RecurrenceConfig copyWith({
    RecurrenceFrequency? frequency,
    int? intervalValue,
    int? maxOccurrences,
    bool clearMaxOccurrences = false,
    DateTime? endDate,
    bool clearEndDate = false,
    ConflictStrategy? conflictStrategy,
  }) {
    return RecurrenceConfig(
      frequency: frequency ?? this.frequency,
      intervalValue: intervalValue ?? this.intervalValue,
      maxOccurrences: clearMaxOccurrences
          ? null
          : (maxOccurrences ?? this.maxOccurrences),
      endDate: clearEndDate ? null : (endDate ?? this.endDate),
      conflictStrategy: conflictStrategy ?? this.conflictStrategy,
    );
  }

  /// Calcola le date delle occorrenze a partire da una data iniziale
  List<DateTime> calculateOccurrences(
    DateTime startDate, {
    int maxPreview = 365,
  }) {
    final dates = <DateTime>[startDate];
    var current = startDate;

    // Se maxOccurrences è specificato, usa quello
    // Altrimenti ("Mai"), calcola fino a 1 anno dalla data iniziale
    final limit = maxOccurrences ?? maxPreview;
    final maxEndDate = endDate ?? startDate.add(const Duration(days: 365));

    while (dates.length < limit) {
      switch (frequency) {
        case RecurrenceFrequency.daily:
        case RecurrenceFrequency.custom:
          current = current.add(Duration(days: intervalValue));
        case RecurrenceFrequency.weekly:
          current = current.add(Duration(days: 7 * intervalValue));
        case RecurrenceFrequency.monthly:
          current = DateTime(
            current.year,
            current.month + intervalValue,
            current.day,
          );
      }

      // Verifica se superato end_date (o limite 1 anno per "Mai")
      if (current.isAfter(maxEndDate)) {
        break;
      }

      dates.add(current);
    }

    return dates;
  }
}

File: agenda_core/src/Domain/Booking/RecurrenceRule.php
<?php

declare(strict_types=1);

namespace Agenda\Domain\Booking;

use DateTimeImmutable;
use JsonSerializable;

/**
 * Regola di ricorrenza per prenotazioni ripetute.
 */
final class RecurrenceRule implements JsonSerializable
{
    public const FREQUENCY_DAILY = 'daily';
    public const FREQUENCY_WEEKLY = 'weekly';
    public const FREQUENCY_MONTHLY = 'monthly';
    public const FREQUENCY_CUSTOM = 'custom';

    public const CONFLICT_SKIP = 'skip';
    public const CONFLICT_FORCE = 'force';

    public function __construct(
        public readonly ?int $id,
        public readonly int $businessId,
        public readonly string $frequency,
        public readonly int $intervalValue,
        public readonly ?int $maxOccurrences,
        public readonly ?DateTimeImmutable $endDate,
        public readonly string $conflictStrategy,
        public readonly ?array $daysOfWeek = null,
        public readonly ?int $dayOfMonth = null,
        public readonly ?DateTimeImmutable $createdAt = null,
        public readonly ?DateTimeImmutable $updatedAt = null,
    ) {}

    /**
     * Crea da array (es. da database row).
     */
    public static function fromArray(array $data): self
    {
        return new self(
            id: isset($data['id']) ? (int) $data['id'] : null,
            businessId: (int) $data['business_id'],
            frequency: $data['frequency'],
            intervalValue: (int) ($data['interval_value'] ?? 1),
            maxOccurrences: isset($data['max_occurrences']) ? (int) $data['max_occurrences'] : null,
            endDate: isset($data['end_date']) ? new DateTimeImmutable($data['end_date']) : null,
            conflictStrategy: $data['conflict_strategy'] ?? self::CONFLICT_SKIP,
            daysOfWeek: isset($data['days_of_week'])
                ? json_decode($data['days_of_week'], true, 512, JSON_INVALID_UTF8_SUBSTITUTE)
                : null,
            dayOfMonth: isset($data['day_of_month']) ? (int) $data['day_of_month'] : null,
            createdAt: isset($data['created_at']) ? new DateTimeImmutable($data['created_at']) : null,
            updatedAt: isset($data['updated_at']) ? new DateTimeImmutable($data['updated_at']) : null,
        );
    }

    /**
     * Calcola tutte le date della serie a partire dalla data iniziale.
     *
     * @return DateTimeImmutable[]
     */
    public function calculateDates(DateTimeImmutable $startDate, int $maxLimit = 52): array
    {
        $dates = [$startDate];
        $current = $startDate;
        $count = 1;

        // Determina limite
        // Se maxOccurrences è specificato, usa quello (con cap a maxLimit per sicurezza)
        // Se non specificato (Mai), usa endDate o calcola fino a 1 anno dalla data iniziale
        if ($this->maxOccurrences !== null) {
            $limit = min($this->maxOccurrences, $maxLimit);
        } else {
            // "Mai" - calcola fino a 1 anno (o endDate se specificata)
            $limit = 365; // Limite massimo di sicurezza per evitare loop infiniti
        }
        
        // Calcola la data massima (1 anno dalla data iniziale se endDate non specificata)
        $maxEndDate = $this->endDate ?? $startDate->modify('+1 year');

        while ($count < $limit) {
            $next = $this->getNextDate($current);

            // Verifica end_date (o limite 1 anno)
            if ($next > $maxEndDate) {
                break;
            }

            $dates[] = $next;
            $current = $next;
            $count++;
        }

        return $dates;
    }

    /**
     * Calcola la prossima data in base alla frequenza.
     */
    private function getNextDate(DateTimeImmutable $current): DateTimeImmutable
    {
        return match ($this->frequency) {
            self::FREQUENCY_DAILY, self::FREQUENCY_CUSTOM => $current->modify("+{$this->intervalValue} days"),
            self::FREQUENCY_WEEKLY => $current->modify("+{$this->intervalValue} weeks"),
            self::FREQUENCY_MONTHLY => $this->getNextMonthlyDate($current),
            default => $current->modify("+{$this->intervalValue} days"),
        };
    }

    /**
     * Calcola la prossima data per ricorrenza mensile.
     * Gestisce il caso in cui il giorno del mese non esiste (es. 31 febbraio).
     */
    private function getNextMonthlyDate(DateTimeImmutable $current): DateTimeImmutable
    {
        $targetDay = $this->dayOfMonth ?? (int) $current->format('j');
        $next = $current->modify("+{$this->intervalValue} months");

        // Se il giorno target è maggiore dei giorni nel mese, usa l'ultimo giorno
        $daysInMonth = (int) $next->format('t');
        $actualDay = min($targetDay, $daysInMonth);

        return $next->setDate(
            (int) $next->format('Y'),
            (int) $next->format('n'),
            $actualDay
        );
    }

    /**
     * Verifica se la strategia è "salta conflitti".
     */
    public function shouldSkipConflicts(): bool
    {
        return $this->conflictStrategy === self::CONFLICT_SKIP;
    }

    /**
     * Verifica se la strategia è "forza creazione".
     */
    public function shouldForceCreation(): bool
    {
        return $this->conflictStrategy === self::CONFLICT_FORCE;
    }

    public function jsonSerialize(): array
    {
        return [
            'id' => $this->id,
            'business_id' => $this->businessId,
            'frequency' => $this->frequency,
            'interval_value' => $this->intervalValue,
            'max_occurrences' => $this->maxOccurrences,
            'end_date' => $this->endDate?->format('Y-m-d'),
            'conflict_strategy' => $this->conflictStrategy,
            'days_of_week' => $this->daysOfWeek,
            'day_of_month' => $this->dayOfMonth,
            'created_at' => $this->createdAt?->format('Y-m-d H:i:s'),
            'updated_at' => $this->updatedAt?->format('Y-m-d H:i:s'),
        ];
    }
}
